Mathlib/Lean/Expr.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2019 Robert Y. Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Simon Hudon, Scott Morrison, Keeley Hoek, Robert Y. Lewis, Floris van Doorn
+-/
+import Lean
+/-!
+# Additional operations on Expr and related types
+
+This file defines basic operations on the types expr, name, declaration, level, environment.
+
+This file is mostly for non-tactics.
+-/
+
+open Lean Meta
+
+namespace Lean
+
+namespace BinderInfo
+
+/-! ### Declarations about `BinderInfo` -/
+
+
+instance : Inhabited BinderInfo :=
+  ⟨BinderInfo.default⟩
+
+/-- The brackets corresponding to a given `BinderInfo`. -/
+def brackets : BinderInfo → String × String
+| BinderInfo.implicit => ("{", "}")
+| BinderInfo.strictImplicit => ("{{", "}}")
+| BinderInfo.instImplicit => ("[", "]")
+| _ => ("(", ")")
+
+end BinderInfo
+
+/-- The type of binders containing a name, the binder info and the binding type -/
+structure Binder :=
+  (name : Name)
+  (info : BinderInfo)
+  (type : Expr)
+deriving Inhabited
+
+namespace Binder
+
+/-- Turn a binder into a string. Uses expr.to_string for the type. -/
+protected def toString (b : Binder) : String :=
+  let (l, r) := b.info.brackets
+  l ++ toString b.name ++ " : " ++ toString b.type ++ r
+
+instance : ToString Binder :=
+⟨Binder.toString⟩
+
+-- instance : ToFormat Binder :=
+-- ⟨λ b => b.toString⟩
+
+-- instance : hasToTacticFormat Binder :=
+--   ⟨fun b =>
+--       let (l, r) := b.info.brackets
+--       (fun e => l ++ b.name.to_string ++ " : " ++ e ++ r) <$> pp b.type⟩
+
+end Binder
+
+namespace Name
+
+/-! ### Declarations about `name` -/
+
+/-- Find the largest prefix `n` of a `Name` such that `f n != none`, then replace this prefix
+with the value of `f n`. -/
+def mapPrefix (f : Name → Option Name) : Name → Name
+| anonymous => anonymous
+| str n' s d => (f (mkStr n' s)).getD (mkStr (mapPrefix f n') s)
+| num n' nr d => (f (mkNum n' nr)).getD (mkNum (mapPrefix f n') nr)
```

Comment:
Some new syntax you might want to check out:
```suggestion
def mapPrefix (f : Name → Option Name) (n : Name) : Name := do
  if let some n' := f n then return n'
  match n with
  | anonymous => anonymous
  | str n' s _ => mkStr (mapPrefix f n') s
  | num n' i _ => mkNum (mapPrefix f n') i
```

I would strongly avoid giving a name to the `UInt64` argument, you can only do bad things with it.